libeos-payg/util.c

```diff
@@ -31,24 +31,28 @@ static gboolean payg_legacy_mode = FALSE;
 /* Force a poweroff in situations where we are not able to enforce PAYG. This
  * is intended to be used as a GSourceFunc, e.g. with g_timeout_add_seconds()
  */
-__attribute__ ((__noreturn__)) gboolean
-payg_sync_and_poweroff (gpointer user_data)
+gboolean
+payg_system_poweroff (gpointer user_data)
 {
-  g_message ("bcsn: Forcing poweroff now!");
-  if (system ("systemctl poweroff") != 0)
+  int ret;
+
+  if (user_data != NULL)
     {
-      /* If we're here, something went wrong in the call, so we can force
-       * an immediate dirty shutdown, since we know a clean shutdown has
-       * already failed.
-       */
-      sync ();
-      reboot (LINUX_REBOOT_CMD_POWER_OFF);
+      guint *source_id = (guint *) user_data;
+      *source_id = 0;
     }
-  /* We've requested a shutdown, so there's nothing left for us to do.
-   * Exit now so the watchdog timer will make sure we get our shutdown
-   * eventually.
-   */
-  exit (EXIT_FAILURE);
+
+  g_message ("Requesting an orderly system shudown");
+  ret = system ("systemctl poweroff");
+  g_debug ("systemctl returned %d", ret);
+
+  /* We have requested an orderly system shutdown. If that request failed of if
+   * a shutdown was already in progress 'systemctl' will return a failure
+   * status, and we can't distinguish between these two scenarios. So let's
```

Comment:
Don’t suppose we could submit a change to `systemctl` to get it to return a different error code if shutdown is already in progress?

Or perhaps we can get the same effect by calling `poweroff` rather than `systemctl poweroff`?

Reply:
On modern systems, `poweroff` is simply a compat entry point for `systemctl poweroff` and has the same return values. We could try that, but I am not convinced that it makes sense to return a success status if the tool failed to request a shutdown because one is already in progress (IDK if systemd / logind even return this info to systemctl).

If we really want to continue forcing a poweroff, we could try to look at `systemctl is-system-running` (which I believe is the same as the `SystemState` property in systemd's D-Bus API) or whether shutdown.target is active, but I'm not sure it is worth since ultimately none of those bits are protected by our security measures.

I have tried to improve the commit message a bit to make this clearer.